src/GraphBLAS-sharp/Implementations.fs

```diff
@@ -115,26 +116,549 @@ type CSRMatrix<'a when 'a : struct and 'a : equality>(csrTuples: CSRFormat<'a>)
+
+    override this.Clear () = failwith "Not Implemented"
+    override this.Copy () = failwith "Not Implemented"
+    override this.Resize a b = failwith "Not Implemented"
+    override this.GetNNZ () = failwith "Not Implemented"
+
+    override this.GetTuples () =
+        opencl {
+            return {| Rows = this.Rows; Columns = this.Columns; Values = this.Values |}
+        }
+
+    override this.GetMask(?isComplemented: bool) =
+        let isComplemented = defaultArg isComplemented false
+        failwith "Not Implemented"
+
+    override this.ToHost () =
+        opencl {
+            let! _ = ToHost this.Rows
+            let! _ = ToHost this.Columns
+            let! _ = ToHost this.Values
+
+            return upcast this
+        }
+
+    override this.Extract (mask: Mask2D option) : OpenCLEvaluation<Matrix<'a>> = failwith "Not Implemented"
+    override this.Extract (colMask: Mask1D option * int) : OpenCLEvaluation<Vector<'a>> = failwith "Not Implemented"
+    override this.Extract (rowMask: int * Mask1D option) : OpenCLEvaluation<Vector<'a>> = failwith "Not Implemented"
+    override this.Extract (idx: int * int) : OpenCLEvaluation<Scalar<'a>> = failwith "Not Implemented"
+    override this.Assign (mask: Mask2D option, value: Matrix<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+    override this.Assign (colMask: Mask1D option * int, value: Vector<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+    override this.Assign (rowMask: int * Mask1D option, value: Vector<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+    override this.Assign (idx: int * int, value: Scalar<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+    override this.Assign (mask: Mask2D option, value: Scalar<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+    override this.Assign (colMask: Mask1D option * int, value: Scalar<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+    override this.Assign (rowMask: int * Mask1D option, value: Scalar<'a>) : OpenCLEvaluation<unit> = failwith "Not Implemented"
+
+    override this.Mxm a b c = failwith "Not Implemented"
+    override this.Mxv a b c = failwith "Not Implemented"
+
+    member internal this.EWiseAddCOO
+        (matrix: COOMatrix<'a>)
+        (mask: Mask2D option)
+        (semiring: Semiring<'a>) : OpenCLEvaluation<Matrix<'a>> =
+
+        let (BinaryOp append) = semiring.PlusMonoid.Append
+        let zero = semiring.PlusMonoid.Zero
+
+        //It is useful to consider that the first array is longer than the second one
+        let firstRows, firstColumns, firstValues, secondRows, secondColumns, secondValues, plus =
+            if this.Rows.Length > matrix.Rows.Length
+            then this.Rows, this.Columns, this.Values, matrix.Rows, matrix.Columns, matrix.Values, append
+            else matrix.Rows, matrix.Columns, matrix.Values, this.Rows, this.Columns, this.Values, <@ fun x y -> (%append) y x @>
```

Comment:
Monoid is commutative in a semiring.

Reply:
I think this variant is better from a technical point of view. Maybe we will change our architecture in future. For example, we might use just monoid that does not require commutability